### PR TITLE
fix: 4.0 table filters value may be previous

### PR DESCRIPTION
**Pull request by jeessy2**
### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

example:
https://codesandbox.io/s/elated-mendeleev-64277
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  4.0 table filters value may be previous when use select, becanse of 'setFilteredKeys' is async  |
| 🇨🇳 Chinese | select选中一个option, 触发的搜索是前一个值 ,因’setFilteredKeys‘是异步的 |

### ☑️ Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by jeessy2** [comment]
@zombieJ added a test case

**Comment by zombieJ** [comment]
Create a hook name `useSyncState` to handle this:
```jsx
const [getFilteredKeys, setFilteredKeys] = React.useState(propFilteredKeys || []);

// ...

const onConfirm = () => {
  internalTriggerFilter(getFilteredKeys());
};
```

**Comment by jeessy2** [comment]
> Create a hook name `useSyncState` to handle this:
> 
> ```js-jsx
> const [getFilteredKeys, setFilteredKeys] = React.useState(propFilteredKeys || []);
> 
> // ...
> 
> const onConfirm = () => {
>   internalTriggerFilter(getFilteredKeys());
> };
> ```
getFilteredKeys(), It's not a function

**Comment by yoyo837** [comment]
```jsx
const [filteredKeys, setFilteredKeys] = React.useState(propFilteredKeys || []);

// ...

const onConfirm = () => {
  internalTriggerFilter(filteredKeys);
};
```

**Comment by jeessy2** [comment]
> ```js-jsx
> const [filteredKeys, setFilteredKeys] = React.useState(propFilteredKeys || []);
> 
> // ...
> 
> const onConfirm = () => {
>   internalTriggerFilter(filteredKeys);
> };
> ```

Use this can't fix the bug

**Comment by yoyo837** [comment]
> getFilteredKeys(), It's not a function

I just correct this code errors.

**Comment by jeessy2** [comment]
create a sync variable to resolve async 'setFilteredKeys'
```
 let filteredKeysLast = filteredKeys;
```

**Comment by zombieJ** [comment]
I'm create a hook your own to handle these code logic:
```jsx
// useSyncState.tsx

export default function useSyncState(...) {
  const stateRef = React.useRef(...);
  // ...
  return [
    () => stateRef.current;
    (newValue) => ...
  ];
}
```

**Comment by jeessy2** [comment]
> I mean to create a hook your own to handle these code logic:
> 
> ```js-jsx
> // useSyncState.tsx
> 
> export default function useSyncState(...) {
>   const stateRef = React.useRef(...);
>   // ...
>   return [
>     () => stateRef.current;
>     (newValue) => ...
>   ];
> }
> ```
> 
> `getXXX` called in `onConfirm` will always get synced state value.

fix to this

File: components/table/hooks/useFilter/FilterDropdown.tsx [hunk 1]
```diff
@@ -92,20 +93,22 @@ function FilterDropdown<RecordType>(props: FilterDropdownProps<RecordType>) {
     filterState && filterState.filteredKeys,
   );
   const [filteredKeys, setFilteredKeys] = React.useState(propFilteredKeys || []);
+  const [getFilteredKeysSync, setFilteredKeysSync] = useSyncState(propFilteredKeys || []);
```
Comment (zombieJ) [inline components/table/hooks/useFilter/FilterDropdown.tsx]: `const [filteredKeys, setFilteredKeys] = React.useState(propFilteredKeys || []);` can be removed now.

File: components/table/hooks/useSyncState.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import * as React from 'react';
+
+type UseSyncStateProps<T> = [() => T, (newValue: T) => void];
+
+export default function useSyncState<T>(filteredKeys: T): UseSyncStateProps<T> {
+  const filteredKeysRef = React.useRef<T>(filteredKeys);
+
+  return [
+    () => filteredKeysRef.current,
+    (newValue: T) => (filteredKeysRef.current = newValue),
```
Comment (zombieJ) [inline components/table/hooks/useSyncState.ts]: Update filter should trigger re-render:
```jsx
const [_, forceUpdate] = React.useState();
// ...
(newValue: T) => {
  filteredKeysRef.current = newValue;
  forceUpdate...
}
```

Comment (zombieJ) [inline components/table/hooks/useSyncState.ts]: Update filter should trigger re-render:
```jsx
const [_, forceUpdate] = React.useState();
// ...
(newValue: T) => {
  filteredKeysRef.current = newValue;
  forceUpdate...
}
```

Comment (zombieJ) [inline components/table/hooks/useSyncState.ts]: Update filter should trigger re-render:
```jsx
const [_, forceUpdate] = React.useState();
// ...
(newValue: T) => {
  filteredKeysRef.current = newValue;
  forceUpdate...
}
```

**Pull request merged**